dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntitySecurityValidationHook.java

```diff
@@ -0,0 +1,122 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.preheat.Preheat;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.security.Authorities;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.preheat.PreheatHelper;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.TrackerIdentifier;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class TrackedEntitySecurityValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 1;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+
+        User actingUser = bundle.getPreheat().getUser();
+
+        for ( TrackedEntity te : bundle.getTrackedEntities() )
+        {
+            reporter.increment();
+
+            TrackedEntityType entityType = getTrackedEntityType( bundle, te );
+
+            if ( entityType != null && !aclService.canDataWrite( actingUser, entityType ) )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1001 )
+                    .withObject( te )
+                    .addArg( actingUser ).addArg( entityType ) );
+            }
+
+            OrganisationUnit orgUnit = getOrganisationUnit( bundle, te );
+
+            if ( orgUnit != null && !organisationUnitService.isInUserSearchHierarchyCached( actingUser, orgUnit ) )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1000 )
+                    .withObject( te )
+                    .addArg( actingUser ).addArg( orgUnit ) );
+            }
+
+            if ( bundle.getImportStrategy().isDelete() )
+            {
+                TrackedEntityInstance tei = PreheatHelper.getTrackedEntityInstance( bundle, te.getTrackedEntity() );
+
+                checkCanCascadeDeleteProgramInstances( reporter, actingUser, tei );
+            }
+        }
+
+        return reporter.getReportList();
+    }
+
+    private void checkCanCascadeDeleteProgramInstances( ValidationErrorReporter errorReporter, User user,
```

Comment:
@Bekkalizer or @mortenoh have to confirm the case of cascade delete, I am not sure about the intentions /details around it.

Reply:
@Bekkalizer @mortenoh ping?

Reply:
lets talk about it in next week meeting, make a note of it